nucliadb/nucliadb/ingest/consumer/auditing.py

```diff
@@ -0,0 +1,295 @@
+from nucliadb.ingest.orm.knowledgebox import KnowledgeBox
+from nucliadb.ingest.orm.nodes_manager import NodesManager
+from nucliadb.ingest.orm.resource import Resource
+from nucliadb_utils import const
+from nucliadb_utils.audit.audit import AuditStorage
+from nucliadb_utils.cache.pubsub import PubSubDriver
+from nucliadb_utils.exceptions import ShardsNotFound
+from nucliadb_utils.storages.storage import Storage
+
+from .utils import DelayedTaskHandler
+
+logger = logging.getLogger(__name__)
+
+
+AUDIT_TYPES = {
+    Notification.WriteType.CREATED: AuditRequest.AuditType.NEW,
+    Notification.WriteType.MODIFIED: AuditRequest.AuditType.MODIFIED,
+    Notification.WriteType.DELETED: AuditRequest.AuditType.DELETED,
+}
+
+
+class IndexAuditHandler:
+    """
+    The purpose of this class is to handle the auditing of data
+    that is indexed on the platform.
+    """
+
+    subscription_id: str
+    loop: asyncio.AbstractEventLoop
+
+    def __init__(
+        self,
+        *,
+        driver: Driver,
+        audit: AuditStorage,
+        pubsub: PubSubDriver,
+        check_delay: float = 5.0,
+    ):
+        self.driver = driver
+        self.audit = audit
+        self.pubsub = pubsub
+        self.node_manager = NodesManager(driver)
+        self.task_handler = DelayedTaskHandler(check_delay)
+
+    async def initialize(self) -> None:
+        self.loop = asyncio.get_running_loop()
+        self.subscription_id = str(uuid.uuid4())
+        await self.task_handler.initialize()
+        await self.pubsub.subscribe(
+            handler=self.handle_message,
+            key=const.PubSubChannels.RESOURCE_NOTIFY.format(kbid="*"),
+            group="audit-shard-stats",
+            subscription_id=self.subscription_id,
+        )
+
+    async def finalize(self) -> None:
+        await self.pubsub.unsubscribe(self.subscription_id)
+        await self.task_handler.finalize()
+
+    async def handle_message(self, raw_data: bytes) -> None:
+        data = self.pubsub.parse(raw_data)
+        notification = Notification()
+        notification.ParseFromString(data)
+
+        if notification.action != Notification.Action.INDEXED:
+            # not a notification we care about
+            return
+
+        self.task_handler.schedule(
+            notification.kbid, partial(self.process_kb, notification.kbid)
+        )
+
+    async def process_kb(self, kbid: str) -> None:
+        try:
+            shard_groups: list[
+                PBShardObject
+            ] = await self.node_manager.get_shards_by_kbid(kbid)
+        except ShardsNotFound:
+            logger.warning(f"No shards found for kbid {kbid}, skipping")
+            return
+
+        total_fields = 0
+        total_paragraphs = 0
+
+        for shard_obj in shard_groups:
+            node, shard_id, _ = self.node_manager.choose_node(shard_obj)
+            shard_counter: Counter = await node.sidecar.GetCount(ShardId(id=shard_id))  # type: ignore
```

Comment:
The `GetCount` logic at the sidecar is caching the counters. May be a good time to review it and see if it makes sense?

https://github.com/nuclia/nucliadb/blob/main/nucliadb_node/nucliadb_node/reader.py#L44

Reply:
ahh nice and then I could just call GetShard on the reader directly and we can remove all that logic

Reply:
I don't know the reasoning behind caching it... maybe @bloodbare or @sunbit know

Reply:
I'm going to leave this alone for now. We'll have to re-evaluate if we want caching at the side car.